compiler/rustc_middle/src/ty/layout.rs

```diff
@@ -1016,9 +1016,16 @@ where
                         tag_field,
                         ..
                     } if this.fields.offset(tag_field) == offset => {
-                        Some(this.for_variant(cx, untagged_variant))
+                        // When a non-null niche is passed in, we might pass an unaligned value.
+                        // Calculates a maximum alignment that matches all niches.
+                        let niche_align = this.variants.niches().map(|niches| {
+                            niches.fold(Align::MAX, |align, niche| {
+                                align.restrict_for_offset(Size::from_bytes(niche))
+                            })
+                        });
+                        (Some(this.for_variant(cx, untagged_variant)), niche_align)
```

Comment:
This logic is still not right, we need something like https://github.com/rust-lang/rust/pull/132745 or else there will be wrong `dereferenceable` flags.

The point of this function is to compute the *pointee* info, but with a type like `MultipleAlign8` in your example, *there is no pointee*.